tests/Jobs/TileSingleOverlayTest.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+namespace Biigle\Tests\Modules\Geo\Jobs;
+
+use Biigle\Modules\Geo\Jobs\TileSingleOverlay;
+use Biigle\Tests\Modules\Geo\GeoOverlayTest;
+use Biigle\FileCache\GenericFile;
+use File;
+use Illuminate\Http\UploadedFile;
+use Illuminate\Support\Facades\Storage;
+use Jcupitt\Vips\Image;
+use Mockery;
+use TestCase;
+
+class TileSingleOverlayTest extends TestCase
+{
+    public function testGenerateOverlayTiles()
+    {
+        Storage::fake('geo-overlays');
+        $overlay = GeoOverlayTest::create();
+
+        // save fake UploadedFile to geo-overlay storage
+        $overlayFile = UploadedFile::fake()->create($overlay->name, 20, 'image/tiff');
+        $overlay->storeFile($overlayFile);
+        $this->assertTrue(Storage::disk('geo-overlays')->exists($overlay->path));
+        
+        // retreive fake UploadedFile from geo-overlay storage and cast to GenericFile
+        $disk = config('geo.tiles.overlay_storage_disk');
+        $file = new GenericFile("{$disk}://{$overlay->path}");
+        
+        $job = new TileSingleOverlayStub($overlay);
+        $mock = Mockery::mock(Image::class);
+        $mock->shouldReceive('dzsave')
+            ->once()
+            ->with($job->tempPath, [
+                'layout' => 'zoomify',
+                'container' => 'fs',
+                'strip' => true,
+            ]);
+
+        $job->mock = $mock;
+
+        $job->generateTiles($file, '');
+    }
+
+    public function testUploadOverlayToStorage()
+    {
+        config(['geo.tiles.overlay_storage_disk' => 'tiles']);
+        $overlay = GeoOverlayTest::create();
+        $fragment = $overlay->id;
+        $job = new TileSingleOverlayStub($overlay);
+        Storage::makeDirectory($job->tempPath);
+        Storage::put("{$job->tempPath}/test.txt", 'test');
+        $this->assertTrue(Storage::disk($job->tempPath)->exists("{$job->tempPath}/test.txt"));
```

Comment:
Hey @mzur, I have trouble understanding where the Storage directory (created with `Storage::makeDirectory()`) is located after creation. I tried to check whether it exists with the assertion above, but I get the error that the disk does not have a configured driver. However, it should have one as it is defined in [TileSingleOverlay](https://github.com/biigle/geo/pull/47/files#diff-b9e6ae069747b38c30ae5c07d197d3e817a35265215faf1eb270be4761f029a2R61) and [geo.php](https://github.com/biigle/geo/pull/47/files#diff-c2d3afafab9f5fe21829b7ad2fc2c8ff1f0a9eefab54410ec5c0e2b5ab4844bcR9), respectively (`'tmp_dir' => sys_get_temp_dir(),`). What do I miss here?

Reply:
You are mixing up local storage and storage disks. The `tmp_dir` is local storage, e.g. `/tmp` where the worker process accumulates the tiles while the job is running. When the job is finished, it uploads all the tiles to the storage disk (which may be a local directory or cloud storage).

In `TileSingleOverlay` you'll notice the difference because it uses `$this->tempPath` for `dzsave()` of Vips. Vips wants to save the files locally and does not know anything about cloud storage. Later in the job in `uploadToStorage()`, it uploads the whole directory of tiles generated by Vips to the storage disk by first getting the disk (`Storage::disk(config('geo.tiles.overlay_storage_disk'))`) and then storing each file (`$disk->putFileAs()`). The argument of `Storage::disk()` is a name of a disk configured in `config/filesystems.php` and not a path as you assumed in your test above.

You can use the [`TileSingleImageTest`](https://github.com/biigle/core/blob/c62ac9eb3c5e9bbc8309b9b25cdb721a484cba84/tests/php/Jobs/TileSingleImageTest.php#L34-L52) as an example for how to test stuff with storage disks. In tests it is important to call [`Storage::fake()`](https://laravel.com/docs/10.x/filesystem#testing) so it takes care of disposing of actual files generated during the test.

Reply:
> The argument of Storage::disk() is a name of a disk configured in config/filesystems.php and not a path as you assumed in your test above.

Thanks, I did not know how to reach the `tempPath` directory once it was created. But I see the mistake I made using the `Storage::disk()` . So I can use `Storage::disk('local')` whenever I want to retrieve something from the local directory, is that correct?

Reply:
Yes and no. These are essentially two different ways to work with files. One is direct local access and the other is an abstraction layer where it does not matter if you work with local files or files in cloud storage. `Storage::disk('local')` is indeed a local directory but you can't access any local file with this, only the files within the specific local directory (configured in `config/filesystems.php`).

When you want to work directly with local files you can use the `File` facade:

```php
use File;

File::put('/tmp/myfile.txt', 'hello world');
```

But why do you need this? Doesn't `TileSingleImage` already implement all the stuff you need?

Reply:
> But why do you need this? Doesn't TileSingleImage already implement all the stuff you need?

Do you mean sth. like using the `TileSingleImage` class (or an extension) also for handling the tiling process for Overlays?

Reply:
> When you want to work directly with local files you can use the File facade

Thank you! Although I've seen this in the `TileSingleImageTest` this didn't make sense to me (I tried to find it in the Laravel documentation but could only find `::makeDirectory`and `::put` in the context of the `Storage` Facade).

Reply:
> Do you mean sth. like using the TileSingleImage class (or an extension) also for handling the tiling process for Overlays?

Yes! You can make TileSingleImage more general so you can extend it for the overlays. This way you don't have to implement everything again.